### PR TITLE
Add equality comparison to c10::Dict

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35019 [jit] kill isSameIdentity
* #35005 [jit] kill `shallowEquals`
* #34986 [jit] define equality for IValue
* #34985 Move ivalue tests to aten
* **#34892 Add equality comparison to c10::Dict**
* #34856 Add equality comparisons to c10::List

Same rationale and implementation as in
https://github.com/pytorch/pytorch/pull/34856

Differential Revision: [D20493169](https://our.internmc.facebook.com/intern/diff/D20493169)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/core/Dict.h [hunk 1]
```diff
@@ -338,6 +338,25 @@ class Dict final {
...
+   * Value equality comparison. This function implements Python-like semantics for
+   * equality: two dicts with the same identity (e.g. same pointer) trivially
+   * compare equal, otherwise each element is compared for equality.
+   */
+  template <class Key_, class Value_>
+  friend bool operator==(
+      const Dict<Key_, Value_>& lhs,
+      const Dict<Key_, Value_>& rhs);
+  template <class Key_, class Value_>
+  friend bool operator!=(
```
Comment (bhosmer) [inline aten/src/ATen/core/Dict.h]: notice `!=` doesn't have an impl or test, maybe that's intended?

**Review by bhosmer — APPROVED** [review]
Couple comments from the peanut gallery FWIW

File: aten/src/ATen/core/Dict.h [hunk 2]
```diff
@@ -338,6 +338,25 @@ class Dict final {
...
+      const Dict<Key_, Value_>& lhs,
+      const Dict<Key_, Value_>& rhs);
+
+  /**
+   * Identity comparison. Returns true if and only if `rhs` represents the same
+   * Dict object as `this`.
+   */
+  bool is(const Dict& is) const;
```
[the remark this answers is not in the document]
Comment (suo) [reply aten/src/ATen/core/Dict.h]: whoops lol

File: aten/src/ATen/core/Dict.h (hunk 1 above)
Comment (suo) [reply aten/src/ATen/core/Dict.h]: good catch! will add

**Pull request closed without merging**